### PR TITLE
Add extraction function to selector dimensional filter to allow lookup on filter

### PR DESCRIPTION
**Pull request by garyluoex**

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/DimensionalFilter.java [hunk 1]
```diff
@@ -14,26 +18,52 @@
...
      *
      * @param dimension  Dimension to filter
      * @param type Type of the filter
      */
     protected DimensionalFilter(Dimension dimension, FilterType type) {
         super(type);
+
+        this.dimension = dimension;
+        if (dimension instanceof ExtractionFunctionDimension) {
+            extractionFunction = ((ExtractionFunctionDimension) dimension).getExtractionFunction().get();
```
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/DimensionalFilter.java]: Are we sure `.get()` returns a extracFn?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java [hunk 2]
```diff
@@ -20,7 +21,19 @@
      * @param value  Value of the filter
      */
     public SelectorFilter(Dimension dimension, String value) {
-        super(dimension, DefaultFilterType.SELECTOR);
+        this(dimension, value, null);
```
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java]: how about calling super that takes 2 args?

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package com.yahoo.bard.webservice.druid.model.filter
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy]: Missing header

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import com.yahoo.bard.webservice.druid.model.dimension.extractionfunction.RegisteredLookupExtractionFunction
+
+import com.fasterxml.jackson.databind.ObjectMapper
+
+import spock.lang.Specification
+
+/**
+ * Test selector filter serialization.
+ */
+class SelectorFilterSpec extends Specification{
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy]: space before `{`

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java [hunk 5]
```diff
@@ -20,7 +21,19 @@
...
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param dimension  Dimension to apply the extraction to
+     * @param value  Value of the filter
+     * @param extractionFn  Extraction function to be applied on dimension
+     *
```
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java]: extra line

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/DimensionalFilter.java (hunk 1 above)
Comment (garyluoex) [reply fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/DimensionalFilter.java]: or null

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java (hunk 2 above)
Comment (garyluoex) [reply fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java]: right

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy [hunk 6]
```diff
@@ -1,3 +1,5 @@
+// Copyright 2017 Yahoo Inc.
```
Comment (QubitPi) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy]: 2018

Comment (garyluoex) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/model/filter/SelectorFilterSpec.groovy]: Yahoo Inc. is not valid also anymore I think

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/ExtractionFilter.java [hunk 7]
```diff
@@ -5,19 +5,18 @@
...
 import com.yahoo.bard.webservice.druid.model.dimension.extractionfunction.ExtractionFunction;
 
-import com.fasterxml.jackson.annotation.JsonProperty;
-
 import java.util.Objects;
 
 /**
  * Filter for matching a dimension using some specific Extraction function.
+ *
+ * @deprecated  Use other dimensional filters with extractionFn specified instead
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/ExtractionFilter.java]: Use javadoc  {@link} to other class.

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java [hunk 8]
```diff
@@ -44,6 +57,17 @@ public SelectorFilter withValue(String value) {
...
+    /**
+     * Get a new instance of this filter with the given value.
+     *
+     * @param extractionFn  Extraction function to be applied on dimension
+     *
+     * @return a new instance of this filter with the given value
+     */
+    public SelectorFilter withExtractionFn(ExtractionFunction extractionFn) {
+        return new SelectorFilter(getDimension(), value, extractionFn);
+    }
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/filter/SelectorFilter.java]: Other withers need to be updated to preserve the extractionFn value.

**Comment by gyehuda** [comment]
Asserting the name _Yahoo Inc._ is the correct copyright holder for code published before Yahoo's name was formally changed in June 2017. Between then and January 2018 you could publish new files using _Yahoo Inc._ or more accurately _Yahoo Holdings Inc_. Either was fine.

Going forward: 
1. no need to change old copyrights on files you are not changing. Think of the copyright statement like a timestamp, not a freshness stamp. Those files will simply say Yahoo Inc. and the year it was published.
2. if you publish a new file now, put a new copyright on it. These days the proper copyright for new files in this project is "Copyright 2018 Oath Inc." 
3. if you are making changes to an existing file, you can, but don't have to update the copyright header. This is one of those inconclusive things that we deal with. We don't use date ranges either. So if the change is minor, don't bother changing the copyright. If it's significant, you can if you wish.

At the end of the day, this code is open source under a permissive license. Everyone in the world can use this code. We simply try to be accurate because we care about quality.

**Pull request merged**